File: JavaScript/DifferentWaysToAddParentheses/DifferentWaysToAddParentheses.js
/**
 * (119ms)
 * @param {string} input
 * @return {number[]}
 */
var diffWaysToCompute = function(input) {
   var ans = [];
   for(var i = 0; i < input.length; i++) {
       if(input[i] === '+' || input[i] === '-' || input[i] === '*') {
           var left = diffWaysToCompute(input.slice(0, i));
           var right = diffWaysToCompute(input.slice(i + 1));
           for(let l of left) {
               for(let r of right) {
                   var res;
                   switch(input[i]) {
                       case '+':
                           res = l + r;
                           break;
                       case '-':
                           res = l - r;
                           break;
                       case '*':
                           res = l * r;
                           break;
                   }
                   ans.push(res);
               }
           }
       }
   }
   if(ans.length === 0) {
       ans.push(Number(input));
   }
   return ans;
};

File: Python/DifferentWaysToAddParentheses/DifferentWaysToAddParentheses.py
from operator import add, sub, mul

class Solution(object):
    # (66ms)
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        op, res = {'+': add, '-': sub, '*': mul}, []
        for i, c in enumerate(input):
            if c in op:
                left, right = self.diffWaysToCompute(input[:i]), self.diffWaysToCompute(input[i+1:])
                res += [op[c](l, r) for l in left for r in right]
        if not res:
            res = [int(input)]
        return res

if __name__ == '__main__':
    print Solution().diffWaysToCompute('2-1-1')
